File: UserService/src/main/scala/APIs/UserService/RegisterUserMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID


/**
 * RegisterUserMessage
 * desc: 根据传入的用户数据创建新用户记录并返回生成的userID。
 * @param username: String (用户的用户名信息。)
 * @param passwordHash: String (用户密码的哈希值。)
 * @param email: String (用户的电子邮箱地址。)
 * @param phoneNumber: String (用户的手机号码。)
 * @return userID: String (生成的新用户ID。)
 */

case class RegisterUserMessage(
  username: String,
  passwordHash: String,
  email: String,
  phoneNumber: String
) extends API[String](UserServiceCode)



case object RegisterUserMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[RegisterUserMessage] = deriveEncoder
  private val circeDecoder: Decoder[RegisterUserMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[RegisterUserMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[RegisterUserMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given registerUserMessageEncoder: Encoder[RegisterUserMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given registerUserMessageDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }


}

File: frontend/src/Plugins/UserService/APIs/RegisterUserMessage.ts
/**
 * RegisterUserMessage
 * desc: 根据传入的用户数据创建新用户记录并返回生成的userID。
 * @param username: String (用户的用户名信息。)
 * @param passwordHash: String (用户密码的哈希值。)
 * @param email: String (用户的电子邮箱地址。)
 * @param phoneNumber: String (用户的手机号码。)
 * @return userID: String (生成的新用户ID。)
 */
import { TongWenMessage } from 'Plugins/TongWenAPI/TongWenMessage'
import { ServiceConfig } from 'Globals/ServiceConfig'



export class RegisterUserMessage extends TongWenMessage {
    constructor(
        public  username: string,
        public  passwordHash: string,
        public  email: string,
        public  phoneNumber: string
    ) {
        super()
    }
    getAddress(): string {
        return ServiceConfig.getUserServiceAddress()
    }
}

